examples/stream/src/main.rs

```diff
@@ -0,0 +1,140 @@
+use std::{io, time::Duration};
+
+use anyhow::{Context, Result};
+use futures::{AsyncReadExt, AsyncWriteExt, StreamExt};
+use libp2p::{multiaddr::Protocol, Multiaddr, PeerId, Stream, StreamProtocol};
+use stream_example as stream;
+use tracing::level_filters::LevelFilter;
+use tracing_subscriber::EnvFilter;
+
+const PROTOCOL: StreamProtocol = StreamProtocol::new("/my-ping-protocol");
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    tracing_subscriber::fmt()
+        .with_env_filter(
+            EnvFilter::builder()
+                .with_default_directive(LevelFilter::INFO.into())
+                .from_env()?,
+        )
+        .init();
+
+    let maybe_address = std::env::args()
+        .nth(1)
+        .map(|arg| arg.parse::<Multiaddr>())
+        .transpose()
+        .context("Failed to parse argument as `Multiaddr`")?;
+
+    let mut swarm = libp2p::SwarmBuilder::with_new_identity()
+        .with_tokio()
+        .with_quic()
+        .with_behaviour(|_| stream::Behaviour::default())?
+        .with_swarm_config(|c| c.with_idle_connection_timeout(Duration::from_secs(10)))
+        .build();
+
+    swarm.listen_on("/ip4/127.0.0.1/udp/0/quic-v1".parse()?)?;
+
+    // Register our custom protocol with the behaviour.
+    let (control, mut incoming_streams) = swarm.behaviour_mut().register(PROTOCOL).unwrap();
+
+    // Deal with incoming streams.
+    // Spawning a dedicated task is just one way of doing this.
+    // libp2p doesn't care how you handle incoming streams but you _must_ handle them somehow.
+    // To mitigate DoS attacks, libp2p will internally drop incoming streams if your application cannot keep in processing them.
```

Comment:
I see two potential mechanisms for handling incoming streams:

1. User provides a handler
2. `stream::Behaviour` provides a `Stream` of streams

Intuitively I opted for (2). The simplicity below also supports (2) to be the better option.

Agreed?

Reply:
Yes. I think (2) is the better API because it gives the user more flexibility.